recipes-core/initrdscripts/files/cryptsetup-imx

```diff
@@ -0,0 +1,97 @@
+# directory will eventually be created by the custom udev rule.
+# This is useful if the rootfs is on a device that takes a while
+# to initialize such as a USB disk.
+wait4file "/dev/disk/by-state" "${bootparam_rootdelay:-0}"
+BALENA_NONENC_BOOT_DEV=""
+
+cryptsetup_enabled() {
+    # Flasher should not try to unlock the partitions
+    if [ "$bootparam_flasher" = "true" ]; then
+        return 1
+    fi
+
+    # Ensure that secure boot is enabled and device is locked
+    if ! is_secured; then
+        return 1
+    fi
+
+    BALENA_NONENC_BOOT_DEV=$(get_state_path_from_label @@BALENA_NONENC_BOOT_LABEL@@)
+    # Only run if the boot partition is split
+    if [ ! -e "$BALENA_NONENC_BOOT_DEV" ]; then
+        return 1
+    fi
+
+    return 0
+}
+
+cryptsetup_run() {
+    TMP_MOUNT_DIR=$(mktemp -d)
+    _tmpkey="ikey"
+    # Configured in the keyctl-caam source
+    _keydir="/tmp/keys"
+    mount "$BALENA_NONENC_BOOT_DEV" "$TMP_MOUNT_DIR"
+
+    for keyfile in "$TMP_MOUNT_DIR"/balena_*.enc; do
+        if [ ! -e "${keyfile}" ]; then
+            fail "No encryption key available"
+        fi
+
+        if ! caam-keygen import "${keyfile}" "${_tmpkey}"; then
+            fail "Failed to import encryption key"
+        fi
+        _keyfile_base=$(basename "${keyfile}")
+        _ksize=${_keyfile_base#balena_}
+        _ksize=${_ksize%.enc}
+    done
+
+    if [ -f ${_keydir}/${_tmpkey} ]; then
+        keyctl padd logon logkey: @u < ${_keydir}/${_tmpkey}
+    else
+        fail "Imported key not found"
+    fi
+
+    BOOT_DEVICE=$(lsblk -nlo pkname "${BALENA_NONENC_BOOT_DEV}")
+    UNLOCKED=""
+    for PART_DEV in $(lsblk -nlo kname,type "/dev/${BOOT_DEVICE}" | grep part | awk '{print $1}'); do
+        PART_NAME=$(lsblk -nlo partlabel "/dev/${PART_DEV}")
+        if [ "${PART_NAME}" = "${BALENA_NONENC_BOOT_LABEL}" ]; then
+            continue
+        fi
+        _valid_partlabels="resin-boot resin-rootA resin-rootB resin-state resin-data"
+        case " ${_valid_partlabels} " in
+            *" ${PART_NAME} "*) ;;
+            *) fail "Trying to mount unknown partition label ${PART_NAME}";;
+        esac
+        _size=$(lsblk -nb -o NAME,SIZE "/dev/${PART_DEV}" | awk '{ print $2 }')
+        _sectors=$( expr "$_size" / 512 )
+        if ! dmsetup -v create "${PART_NAME}" --table "0 $_sectors crypt capi:tk(cbc(aes))-plain :$_ksize:logon:logkey: 0 /dev/${PART_DEV} 0 1 sector_size:512"; then
+            fail "Failed to unlock encrypted partition ${PART_DEV}"
+        fi
+        UNLOCKED="${UNLOCKED} ${PART_NAME}"
+    done
+
+    # Wait for udev processing of each unlocked device
+    for DM_NAME in ${UNLOCKED}; do
+        wait4udev "/dev/mapper/${DM_NAME}"
+    done
+
+    # udev rules run after dmsetup but there is a race condition between
+    # dm rules and the balena state rules that results in state links
+    # not being created. Needs more investigation, but for the time being
+    # re-run them here.
+    udevadm trigger
```

Comment:
I suspect this is the same as on x86. The balena state rules only create by-state symlinks for partitions on the same device as the root partition, but we don't know where the root partition is until it gets decrypted. So anything unlocked before the root partition will lack the symlink.

Reply:
But on x86 we don't retrigger udev, do we? I remember comparing with the x86 implementation and cryptsetup does seem to make this work.

Reply:
Not in the `cryptsetup` script, but IIRC all the following scripts that need to use the links trigger & settle themselves. Not intentionally though, they just happen to do that. Do you know what exactly breaks on the iMX8? The scripts should not be that different between device types.

Reply:
Nope, I am convinced it's a race condition with the `dm-` rules, but I failed to fix by modifying the rules themselves (sleeps etc) and I didn't want to spend to much time on it as the re-trigger worked.

My intuition tells me that with LUKS the rules do not race against each other.